src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,129 @@
+    private final TagDAO tagDAO;
+    private final OrderDAO orderDAO;
+    private final MapperDTO mapperDTO;
+
+    @Override
+    @Transactional
+    public CertificateDTO create(CertificateDTO certificateDTO) {
+        Certificate certificate = mapperDTO.convertDTOToCertificate(certificateDTO);
+        certificate.setTags(new HashSet<>());
+        attachTags(certificate, certificateDTO.getTags());
+        certificate = certificateDAO.create(certificate);
+        return mapperDTO.convertCertificateToDTO(certificate);
+    }
+
+    @Override
+    public List<CertificateDTO> findAll(Page page) {
+        return certificateDAO.findAll(page)
+            .stream()
+            .map(mapperDTO::convertCertificateToDTO)
+            .collect(Collectors.toList());
+    }
+
+    @Override
+    public CertificateDTO findById(Long id) {
+        Certificate certificate = certificateDAO.findById(id)
+            .orElseThrow(() -> new CertificateNotFoundException(id.toString()));
+        if (!certificate.isActive()) {
+            throw new CertificateNotFoundException(id.toString());
+        }
+        return mapperDTO.convertCertificateToDTO(certificate);
+    }
+
+    @Override
+    public List<CertificateDTO> findAll(ParametersSpecificationDTO querySpecificationDTO,
+                                        Page page) {
+        QuerySpecification querySpecification = mapperDTO.convertDTOToQuery(querySpecificationDTO);
+        return certificateDAO.findAll(querySpecification, page)
+            .stream()
+            .map(mapperDTO::convertCertificateToDTO)
+            .collect(Collectors.toList());
+    }
+
+    @Override
+    @Transactional
+    public CertificateDTO update(CertificateDTO updateDTO) {
+        Optional<Certificate> optional = certificateDAO.findById(updateDTO.getId());
+        if (optional.isEmpty() || !optional.get().isActive()) {
+            throw new CertificateNotFoundException(updateDTO.getId().toString());
+        }
+        Certificate certificate = optional.get();
+        Certificate update = mapperDTO.convertDTOToCertificate(updateDTO);
+        certificate = certificateDAO.update(certificate, update);
+        attachTags(certificate, updateDTO.getTags());
+        return mapperDTO.convertCertificateToDTO(certificate);
+    }
+
+    @Override
+    @Transactional
+    public CertificateDTO applyPatch(Long id, PatchDTO patchDTO) {
+        Certificate certificate = certificateDAO.findById(id)
+            .orElseThrow(() -> new CertificateNotFoundException(id.toString()));
+        if (!certificate.isActive()) {
+            throw new CertificateNotFoundException(id.toString());
+        }
+        Certificate update = mapperDTO.convertPatchDTOToCertificate(patchDTO);
+        certificate = certificateDAO.applyPatch(certificate, update);
+        return mapperDTO.convertCertificateToDTO(certificate);
+    }
+
+    @Override
+    public List<CertificateDTO> findAllByOrderId(Long id, Page page) {
```

Comment:
here the pagination is not needed

Reply:
A person can have a lot of orders.
The role of a person can be performed by a company or other legal entity. I suggest leaving pagination.

Reply:
> 

A person can have a lot of orders.
The role of a person can be performed by a company or other legal entity. I suggest leaving pagination.